### ISSUE TITLE
[ENH] Validation loss for deep learning clustering, classification and regression

### ISSUE THREAD

**Issue by PipaFlores** [aeon-toolkit/aeon#3039]
### Describe the feature or idea you want to propose

I am testing the different AutoEncoders implemented in aeon. I have noticed that the models are only storing the training loss, accessible for plotting/inspection in `model.summary()`. However, there are no validation losses. 

### Describe your proposed solution

I made a small addition in the `self.training_model_.fit()` in 

https://github.com/aeon-toolkit/aeon/blob/0412d5b50272f58efbcd989ffe18bb46d6378965/aeon/clustering/deep_learning/_ae_abgru.py#L284-L292

where only a new argument of `validation_split = int [0,1] ` is needed according to keras api https://keras.io/api/models/model_training_apis/

I did this on my experiments and it seems to work fine. I would consider this enhancement very useful, as naive users might go for the 2000 n_epochs default run and end up overfitting all their models, and then the manifold learning and subsequent clustering makes no much sense 

I would be happy to create a PR for most, if not all, the models in the deep learning clustering module. 
Basically, it would be a new argument for the initialization of the clusterer classes (validation_split).

I am new to Keras, so please let me know if I overlooked any easier way to do this.

**Comment by PipaFlores** [comment]
Another open question on my side is on how the `save_best_model` works. Its creating some `callbacks.modelCheckpoint(..., monitor = "loss", save_best_only = True)`
So the best model is defined as the one with min(loss)? I would switch this to a validation loss criteria instead, as it is overlooking the issue of overfitting.

**Comment by hadifawaz1999** [comment]
Hello, thank you for raising the enhancement proposal. We've been asked a lot before if all deep learning models do support validation split, in classification regression and clustering, its always been on our to do list but still havent done it, am happy to have this functionality, although to make it clear because you also mentioned the save_best_model thing:

We follow this rule: a deep learning model can be trained to infinite amount of epochs, making it random if we use the "last model" for evaluation in predict. In practice we use the "best model" for evaluation. However the "best model" can be chosen either on the validation loss if a validation split is provided and on the training loss if not. By default in aeon the ModelCheckpoint saves to file the best model using the training loss. Now if ```save_best_model``` is set to true, the keras file will be kept in file after fit is done, if its false then the file will be deleted (see docs of any deep model in aeon). The saving directory is the ```file_path``` parameter and the name of the file will be ```best_file_name``` parameter with a default added ```.keras``` extension.

So if you want to add the validation spit parameter am more than happy to help review the PR however the following should be done:

- A new parameter called ```validation_split``` default to ```0.0```, explained in docs similar to its docs in keras (see [here](https://keras.io/api/models/model_training_apis/))
- Mention in the docs that if validation_split is > 0.0 then the "val_loss" is used in modelcheckpoint
- In the code where modelCheckpoint is being called, there should be an if else statement, if ```validation_split``` is zero then use "loss" to "monitor" and if its >0.0 then use "val_loss" to "monitor" with the rest of the functionality kept the same

Also there should be a new test just for this functionality to make sure its working, under aeon/clustering/deep_learning/tests with a new file called test_deep_clustering_validation.py

**Comment by hadifawaz1999** [comment]
Am happy to get this in for now into clustering only, and once we have it and ***tested*** then we get this in for classification and regression

**Comment by jsquaredosquared** [comment]
> where only a new argument of `validation_split = int [0,1] ` is needed according to keras api https://keras.io/api/models/model_training_apis/

Can y'all consider adding the `validation_data` parameter as well? Sometimes it is helpful to be able to pass validation data from a predefined and performed split. According to the link, this parameter overrides `validation_split`.

**Comment by hadifawaz1999** [comment]
> > where only a new argument of `validation_split = int [0,1] ` is needed according to keras api https://keras.io/api/models/model_training_apis/
> 
> Can y'all consider adding the `validation_data` parameter as well? Sometimes it is helpful to be able to pass validation data from a predefined and performed split. According to the link, this parameter overrides `validation_split`.

it would be a good addition, do you want to make a PR for it ? Given we still haven't done this for classification/regression but only for classification you can add both validation_split and validation_data

**Comment by jsquaredosquared** [comment]
Hi @hadifawaz1999, I'd love to take a crack at it. I think we can expand to use the full compile and fit model training APIs in one fell swoop. Here is my initial idea:

## Suggested changes
We could add 2 new parameters to the constructors of the deep learning estimators, `compile_args: dict` and `fit_args: dict`. This will avoid further ballooning of the argument list if each parameter were added separately, and it would only require 2 new entries in the documentation. The docs could state that these dictionaries house Keras compile and fit parameters that are not already handled by Aeon's deep learning constructors, and a link to the Keras docs would be provided.


Then, when Aeon is ready to invoke the model compile and fit methods, we unpack these provided arguments. This will allow the user to make use of the full Keras model training API for the compile and fit methods. A valid invocation might look like this:
```python
clf = ResNetClassifier(
    n_epochs=20,
    batch_size=4,
    compile_args={"weighted_metrics": ["accuracy"]},
    fit_args={
        "validation_data": (x_val, y_val),
        "class_weight": {
            0: 1.0,
            1: 5.0
        }
    },
)
```
Then when the model is compiled and fitted:
```python
model = tf.keras.models.Model(inputs=input_layer, outputs=output_layer)
model.compile(
    loss=self.loss,
    optimizer=self.optimizer_,
    metrics=self._metrics,
    **self.compile_args,
)
```

```python
self.history = self.training_model_.fit(
    X,
    y_onehot,
    batch_size=mini_batch_size,
    epochs=self.n_epochs,
    verbose=self.verbose,
    callbacks=self.callbacks_,
    **self.fit_args,
)
```

## Potenital errors
- If the user specifies a parameter in one of these dicts that is already handled by Aeon's constructors, a `TypeError` would be thrown (e.g., `got multiple values for keyword argument 'callbacks'`). We could warn the user in the docs not to provide these args in the dicts if they are already handled by the constructor, or we could add a check in the code.

- If the user specifies an invalid parameter that Keras does not accept, a `TypeError` would be thrown (e.g., `got an unexpected keyword argument 'validation_frequency'`). The user would need to spell the keys correctly, and only provide parameters that are accepted by Keras's API, with the correct values. A typed dict could also be implemented to help with providing the correct keys and value types.

- Using this unpacking method also means that if an argument is not provided, Keras's default value will just be used anyway.

---
I believe this change should work for all the DL-based estimators. What do you think?

**Comment by Shreyas1627** [comment]
"Hi @PipaFlores , I'm Shreyas, a CS undergrad with experience in TensorFlow/Keras (CNNs) and FastAPI. I'm very interested in helping implement validation loss for the deep learning estimators as part of my GSoC 2026 preparation. Is there a specific estimator (e.g., classification or clustering) you'd like me to start with? I'd love to take this up!"
